algoliasearch-common/src/main/java/com/algolia/search/AsyncIndexCRUD.java

```diff
@@ -122,4 +129,115 @@
+  /**
+   * Rebuild the entirety of your data atomically, to ensure that searches performed on the index
+   * during the rebuild will not be interrupted.
+   *
+   * @param indexContent the content of the index to reindex
+   */
+  default List<Long> reIndex(@Nonnull IndexContent<T> indexContent)
+      throws InterruptedException, ExecutionException, AlgoliaException {
+    return reIndex(indexContent, false, new RequestOptions());
+  }
+
+  /**
+   * Rebuild the entirety of your data atomically, to ensure that searches performed on the index
+   * during the rebuild will not be interrupted.
+   *
+   * @param indexContent the content of the index to reindex
+   * @param safe runs the reindex safely
+   */
+  default List<Long> reIndex(@Nonnull IndexContent<T> indexContent, boolean safe)
+          throws InterruptedException, ExecutionException, AlgoliaException {
+    return reIndex(indexContent, safe, new RequestOptions());
+  }
+
+  /**
+   * Rebuild the entirety of your data atomically, to ensure that searches performed on the index
+   * during the rebuild will not be interrupted.
+   *
+   * @param indexContent the content of the index to reindex
+   * @param safe runs the reindex safely
+   * @param requestOptions request options
+   */
+  default List<Long> reIndex(
+      @Nonnull IndexContent<T> indexContent, boolean safe, @Nonnull RequestOptions requestOptions)
+      throws ExecutionException, InterruptedException, AlgoliaException {
+
+    // 1. Init temps Index
+    AsyncIndex<T> tmpIndex =
+        getApiClient()
+            .initIndex(getName() + "_tmp_" + UUID.randomUUID().toString(), indexContent.getObjectClass());
+
+    // 2. Copy the settings, synonyms and rules (but not the records)
+    List<Long> taskIds = new ArrayList<>();
+    List<String> scopes = indexContent.getScopes();
+
+    // Copy index
+    if (scopes.size() > 0) {
+      AsyncTask copyIndexTask = copyTo(tmpIndex.getName(), scopes, requestOptions).get();
+      taskIds.add(copyIndexTask.getTaskID());
+      if (safe) {
+        tmpIndex.waitTask(copyIndexTask.getTaskID());
+      }
+    }
+
+    // Rules
+    if (indexContent.getRules() != null) {
+      AsyncTask task = tmpIndex.batchRules(indexContent.getRules(), requestOptions).get();
+      taskIds.add(task.getTaskID());
+    }
+
+    // Settings
+    if (indexContent.getSettings() != null) {
+      AsyncTask task = tmpIndex.setSettings(indexContent.getSettings(), requestOptions).get();
+      taskIds.add(task.getTaskID());
+    }
+
+    // Synonyms
+    if (indexContent.getSynonyms() != null) {
+      AsyncTask task = tmpIndex.batchSynonyms(indexContent.getSynonyms(), requestOptions).get();
+      taskIds.add(task.getTaskID());
+    }
+
+    // 3. Fetch your data with the iterator and push it to the temporary index
+    if (indexContent.getObjects() != null) {
+      ArrayList<T> records = new ArrayList<>();
+
+      for (T object : indexContent.getObjects()) {
+
+        if (records.size() == 1000) {
```

Comment:
Performance issue ( at least on PHP ): You may want to increase a counter instead of calculating the size of the records every object.